diff --git a/README.mlton b/README.mlton
new file mode 100644
index 0000000..f1b9b15
--- /dev/null
+++ b/README.mlton
@@ -0,0 +1,8 @@
+The following changes were made to the ckit Library, in addition to
+deriving the `.mlb` file from the `.cm` file:
+
+* `ast/pp/pp-ast-adornment-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `ast/pp/pp-ast-ext-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `ast/type-util-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `parser/parse-tree-sig.sml` (modified): Rewrote use of (sequential) `withtype` in signature.
+* `parser/parse-tree.sml` (modified): Rewrote use of (sequential) `withtype`.
diff --git a/ckit-lib.mlb b/ckit-lib.mlb
new file mode 100644
index 0000000..abc601f
--- /dev/null
+++ b/ckit-lib.mlb
@@ -0,0 +1 @@
+src/ckit-lib.mlb
diff --git a/src/ast/pp/pp-ast-adornment-sig.sml b/src/ast/pp/pp-ast-adornment-sig.sml
index a7b937b..ab6e0a9 100644
--- a/src/ast/pp/pp-ast-adornment-sig.sml
+++ b/src/ast/pp/pp-ast-adornment-sig.sml
@@ -1,14 +1,14 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a pp =  Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
 
   type ('aidinfo,'a,'b) adornment_pp = ('aidinfo -> 'a) -> 'aidinfo -> 'b
-in
+(* in *)
 signature PPASTADORNMENT = sig
   type aidinfo
   val ppExpressionAdornment: (aidinfo,Ast.coreExpression pp,Ast.expression pp) adornment_pp
   val ppStatementAdornment : (aidinfo,Ast.coreStatement pp,Ast.statement pp) adornment_pp
   val ppExternalDeclAdornment: (aidinfo,Ast.coreExternalDecl pp,Ast.externalDecl pp) adornment_pp
 end
-end
+(* end *)
diff --git a/src/ast/pp/pp-ast-ext-sig.sml b/src/ast/pp/pp-ast-ext-sig.sml
index 4169fc7..c7291ab 100644
--- a/src/ast/pp/pp-ast-ext-sig.sml
+++ b/src/ast/pp/pp-ast-ext-sig.sml
@@ -1,13 +1,13 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a pp =  Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
   type ('a, 'aidinfo) ppExt =
        (('aidinfo -> Ast.expression pp) * ('aidinfo -> Ast.statement pp) *
 	('aidinfo -> Ast.binop pp) * ('aidinfo -> Ast.unop pp))
        -> 'aidinfo
        -> Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
-in
+(* in *)
 
 signature PPASTEXT = sig
   type aidinfo
@@ -24,4 +24,4 @@ signature PPASTEXT = sig
        aidinfo) ppExt
 end
 
-end
+(* end *)
diff --git a/src/ast/type-util-sig.sml b/src/ast/type-util-sig.sml
index b03260b..a262146 100644
--- a/src/ast/type-util-sig.sml
+++ b/src/ast/type-util-sig.sml
@@ -1,10 +1,10 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a type_util      = Tables.tidtab -> Ast.ctype               -> 'a 
   type 'a type_mem_util  = Tables.tidtab -> Ast.ctype * Ast.member  -> 'a 
   type 'a type_type_util = Tables.tidtab -> Ast.ctype * Ast.ctype -> 'a 
-in
+(* in *)
 
 signature TYPE_UTIL =
 sig
@@ -145,4 +145,4 @@ sig
 
 end (* signature TYPE_UTIL *)
 
-end (* local *)
+(* end (* local *) *)
diff --git a/src/ckit-lib.mlb b/src/ckit-lib.mlb
new file mode 100644
index 0000000..04c7479
--- /dev/null
+++ b/src/ckit-lib.mlb
@@ -0,0 +1,890 @@
+
+ann
+  "nonexhaustiveBind warn" "nonexhaustiveMatch warn"
+  "redundantBind warn" "redundantMatch warn"
+  "sequenceNonUnit ignore"
+  "warnUnused false" "forceUsed"
+  "allowOrPats true" "allowSigWithtype true" "allowVectorExpsAndPats true"
+in
+
+local
+  basis l4 =
+    bas
+      (* $/basis.cm ====> *) $(SML_LIB)/basis/basis.mlb $(SML_LIB)/basis/unsafe.mlb
+    end
+  basis l24 =
+    bas
+      (* $/smlnj-lib.cm ====> *) $(SML_LIB)/smlnj-lib/Util/smlnj-lib.mlb
+    end
+  basis l71 =
+    bas
+      (* $/pp-lib.cm ====> *) $(SML_LIB)/smlnj-lib/PP/pp-lib.mlb
+    end
+  basis l96 =
+    bas
+      (* $/ml-yacc-lib.cm ====> *) $(SML_LIB)/mlyacc-lib/mlyacc-lib.mlb
+    end
+in
+local
+   $(SML_LIB)/basis/pervasive.mlb
+   local
+      open l4
+   in
+      structure gs_0 = TextIO
+   end
+   local
+      variants/type-check-control-sig.sml
+   in
+      signature gs_1 = TYPECHECKCONTROL
+   end
+   local
+      variants/parse-control-sig.sml
+   in
+      signature gs_2 = PARSECONTROL
+   end
+   local
+      signature PARSECONTROL = gs_2
+      signature TYPECHECKCONTROL = gs_1
+      variants/config-sig.sml
+   in
+      signature gs_3 = CONFIG
+   end
+   local
+      signature CONFIG = gs_3
+      signature PARSECONTROL = gs_2
+      signature TYPECHECKCONTROL = gs_1
+      structure TextIO = gs_0
+      variants/ansic/config.sml
+   in
+      structure gs_4 = Config
+   end
+   local
+      open l24
+   in
+      functor gs_5 = HashTableFn
+   end
+   local
+      ast/uidtabimp-sig.sml
+   in
+      signature gs_6 = UIDTABIMP
+   end
+   local
+      open l4
+   in
+      structure gs_7 = Word
+   end
+   local
+      structure Word = gs_7
+      ast/uid-sig.sml
+   in
+      signature gs_8 = UID
+   end
+   local
+      functor HashTableFn = gs_5
+      signature UID = gs_8
+      signature UIDTABIMP = gs_6
+      ast/uidtabimp-fn.sml
+   in
+      functor gs_9 = UidtabImpFn
+   end
+   local
+      open l4
+   in
+      structure gs_10 = Int
+   end
+   local
+      structure Int = gs_10
+      signature UID = gs_8
+      structure Word = gs_7
+      ast/uid-fn.sml
+   in
+      functor gs_11 = UidFn
+   end
+   local
+      signature UID = gs_8
+      functor UidFn = gs_11
+      ast/aid.sml
+   in
+      structure gs_12 = Aid
+   end
+   local
+      structure Aid = gs_12
+      functor UidtabImpFn = gs_9
+      ast/aidtab.sml
+   in
+      structure gs_13 = Aidtab
+   end
+   local
+      open l24
+   in
+      structure gs_14 = Format
+   end
+   local
+      open l4
+   in
+      structure gs_15 = String
+   end
+   local
+      parser/util/sourcemap-sig.sml
+   in
+      signature gs_16 = SOURCE_MAP
+   end
+   local
+      structure Config = gs_4
+      structure Format = gs_14
+      structure Int = gs_10
+      signature SOURCE_MAP = gs_16
+      structure String = gs_15
+      parser/util/sourcemap.sml
+   in
+      structure gs_17 = SourceMap
+   end
+   local
+      open l71
+   in
+      functor gs_18 = PPStreamFn
+   end
+   local
+      open l71
+   in
+      structure gs_19 = StringToken
+   end
+   local
+      open l4
+   in
+      structure gs_20 = StringCvt
+   end
+   local
+      open l4
+   in
+      structure gs_21 = List
+   end
+   local
+      structure List = gs_21
+      functor PPStreamFn = gs_18
+      structure String = gs_15
+      structure StringCvt = gs_20
+      structure StringToken = gs_19
+      parser/util/old-pp.sml
+   in
+      signature gs_22 = OLD_PRETTYPRINT
+      structure gs_23 = OldPrettyPrint
+   end
+   local
+      structure Format = gs_14
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure SourceMap = gs_17
+      structure TextIO = gs_0
+      parser/util/error-sig.sml
+   in
+      signature gs_24 = ERROR
+   end
+   local
+      signature ERROR = gs_24
+      structure Format = gs_14
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure SourceMap = gs_17
+      structure TextIO = gs_0
+      parser/util/error.sml
+   in
+      structure gs_25 = Error
+   end
+   local
+      open l96
+   in
+      functor gs_26 = Join
+      functor gs_27 = JoinWithArg
+   end
+   local
+      open l96
+   in
+      structure gs_28 = LrParser
+   end
+   local
+      open l4
+   in
+      structure gs_29 = LargeInt
+   end
+   local
+      parser/extensions/c/parse-tree-ext.sml
+   in
+      structure gs_30 = ParseTreeExt
+   end
+   local
+      structure LargeInt = gs_29
+      structure ParseTreeExt = gs_30
+      structure SourceMap = gs_17
+      parser/parse-tree-sig.sml
+   in
+      signature gs_31 = PARSETREE
+   end
+   local
+      structure LargeInt = gs_29
+      signature PARSETREE = gs_31
+      structure ParseTreeExt = gs_30
+      structure SourceMap = gs_17
+      parser/parse-tree.sml
+   in
+      structure gs_32 = ParseTree
+   end
+   local
+      structure Error = gs_25
+      structure ParseTree = gs_32
+      parser/parser-sig.sml
+   in
+      signature gs_33 = PARSER
+   end
+   local
+      open l4
+   in
+      structure gs_34 = Unsafe
+   end
+   local
+      open l4
+   in
+      structure gs_35 = IntInf
+   end
+   local
+      open l4
+   in
+      structure gs_36 = CharVector
+   end
+   local
+      open l4
+   in
+      structure gs_37 = Vector
+   end
+   local
+      open l4
+   in
+      structure gs_38 = Real
+   end
+   local
+      open l4
+   in
+      structure gs_39 = Char
+   end
+   local
+      open l24
+   in
+      structure gs_40 = AtomTable
+   end
+   local
+      open l24
+   in
+      structure gs_41 = Atom
+   end
+   local
+      structure Atom = gs_41
+      structure AtomTable = gs_40
+      structure Config = gs_4
+      parser/grammar/tdefs.sml
+   in
+      signature gs_42 = TYPEDEFS
+      structure gs_43 = TypeDefs
+   end
+   local
+      open l96
+   in
+      signature gs_44 = ARG_LEXER
+      signature gs_45 = ARG_PARSER
+      signature gs_46 = LEXER
+      signature gs_47 = LR_PARSER
+      signature gs_48 = LR_TABLE
+      signature gs_49 = PARSER
+      signature gs_50 = PARSER_DATA
+      signature gs_51 = STREAM
+      signature gs_52 = TOKEN
+   end
+   local
+      signature ARG_LEXER = gs_44
+      signature ARG_PARSER = gs_45
+      signature LEXER = gs_46
+      signature LR_PARSER = gs_47
+      signature LR_TABLE = gs_48
+      structure LargeInt = gs_29
+      signature PARSER = gs_49
+      signature PARSER_DATA = gs_50
+      signature STREAM = gs_51
+      signature TOKEN = gs_52
+      parser/grammar/c.grm.sig
+   in
+      signature gs_53 = C_LRVALS
+      signature gs_54 = C_TOKENS
+   end
+   local
+      structure Atom = gs_41
+      structure AtomTable = gs_40
+      signature C_LRVALS = gs_53
+      signature C_TOKENS = gs_54
+      structure Config = gs_4
+      signature TYPEDEFS = gs_42
+      structure TypeDefs = gs_43
+      parser/grammar/tokentable.sml
+   in
+      signature gs_55 = TOKENTABLE
+      functor gs_56 = TokenTable
+   end
+   local
+      signature C_LRVALS = gs_53
+      signature C_TOKENS = gs_54
+      structure Char = gs_39
+      structure CharVector = gs_36
+      structure Int = gs_10
+      structure IntInf = gs_35
+      structure LargeInt = gs_29
+      structure List = gs_21
+      structure Real = gs_38
+      structure SourceMap = gs_17
+      structure String = gs_15
+      structure StringCvt = gs_20
+      signature TOKENTABLE = gs_55
+      functor TokenTable = gs_56
+      structure Unsafe = gs_34
+      structure Vector = gs_37
+      parser/grammar/c.lex.sml
+   in
+      functor gs_57 = CLexFun
+   end
+   local
+      open l4
+   in
+      structure gs_58 = Array
+   end
+   local
+      open l4
+   in
+      structure gs_59 = General
+   end
+   local
+      signature ARG_LEXER = gs_44
+      signature ARG_PARSER = gs_45
+      structure Array = gs_58
+      signature C_LRVALS = gs_53
+      signature C_TOKENS = gs_54
+      structure Char = gs_39
+      structure Error = gs_25
+      structure General = gs_59
+      signature LEXER = gs_46
+      signature LR_PARSER = gs_47
+      signature LR_TABLE = gs_48
+      structure LargeInt = gs_29
+      structure List = gs_21
+      signature PARSER = gs_49
+      signature PARSER_DATA = gs_50
+      structure ParseTree = gs_32
+      signature STREAM = gs_51
+      structure SourceMap = gs_17
+      structure String = gs_15
+      signature TOKEN = gs_52
+      signature TYPEDEFS = gs_42
+      structure TypeDefs = gs_43
+      parser/grammar/c.grm.sml
+   in
+      functor gs_60 = LrValsFun
+   end
+   local
+      functor CLexFun = gs_57
+      structure Error = gs_25
+      functor Join = gs_26
+      functor JoinWithArg = gs_27
+      structure LrParser = gs_28
+      functor LrValsFun = gs_60
+      signature PARSER = gs_33
+      structure SourceMap = gs_17
+      signature TOKENTABLE = gs_55
+      signature TYPEDEFS = gs_42
+      structure TextIO = gs_0
+      functor TokenTable = gs_56
+      structure TypeDefs = gs_43
+      parser/parser.sml
+   in
+      structure gs_61 = Parser
+   end
+   local
+      open l24
+   in
+      structure gs_62 = HashString
+   end
+   local
+      signature UID = gs_8
+      functor UidFn = gs_11
+      ast/tid.sml
+   in
+      structure gs_63 = Tid
+   end
+   local
+      structure Tid = gs_63
+      ast/symbol-sig.sml
+   in
+      signature gs_64 = SYMBOL
+   end
+   local
+      structure HashString = gs_62
+      structure Int = gs_10
+      signature SYMBOL = gs_64
+      structure String = gs_15
+      structure Tid = gs_63
+      structure Word = gs_7
+      ast/symbol.sml
+   in
+      structure gs_65 = Symbol
+   end
+   local
+      signature UID = gs_8
+      functor UidFn = gs_11
+      ast/pid.sml
+   in
+      structure gs_66 = Pid
+   end
+   local
+      ast/extensions/c/ast-ext.sml
+   in
+      structure gs_67 = AstExt
+   end
+   local
+      structure Aid = gs_12
+      structure AstExt = gs_67
+      structure LargeInt = gs_29
+      structure Pid = gs_66
+      structure SourceMap = gs_17
+      structure Symbol = gs_65
+      structure Tid = gs_63
+      ast/ast-sig.sml
+   in
+      signature gs_68 = AST
+   end
+   local
+      signature AST = gs_68
+      structure Aid = gs_12
+      structure AstExt = gs_67
+      structure LargeInt = gs_29
+      structure Pid = gs_66
+      structure SourceMap = gs_17
+      structure Symbol = gs_65
+      structure Tid = gs_63
+      ast/ast.sml
+   in
+      structure gs_69 = Ast
+   end
+   local
+      structure Ast = gs_69
+      structure LargeInt = gs_29
+      structure Pid = gs_66
+      structure SourceMap = gs_17
+      structure Symbol = gs_65
+      structure Tid = gs_63
+      ast/bindings.sml
+   in
+      structure gs_70 = Bindings
+   end
+   local
+      open l24
+   in
+      functor gs_71 = BinaryMapFn
+   end
+   local
+      open l24
+   in
+      signature gs_72 = ORD_MAP
+   end
+   local
+      structure Tid = gs_63
+      functor UidtabImpFn = gs_9
+      ast/tidtab.sml
+   in
+      structure gs_73 = Tidtab
+   end
+   local
+      structure Aidtab = gs_13
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Tidtab = gs_73
+      ast/tables.sml
+   in
+      structure gs_74 = Tables
+   end
+   local
+      structure Aid = gs_12
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Error = gs_25
+      structure LargeInt = gs_29
+      signature ORD_MAP = gs_72
+      structure SourceMap = gs_17
+      structure Symbol = gs_65
+      structure Tables = gs_74
+      structure Tid = gs_63
+      ast/state-sig.sml
+   in
+      signature gs_75 = STATE
+   end
+   local
+      structure Aid = gs_12
+      structure Aidtab = gs_13
+      structure Ast = gs_69
+      functor BinaryMapFn = gs_71
+      structure Bindings = gs_70
+      structure Error = gs_25
+      structure LargeInt = gs_29
+      structure List = gs_21
+      structure Pid = gs_66
+      signature STATE = gs_75
+      structure SourceMap = gs_17
+      structure Symbol = gs_65
+      structure Tables = gs_74
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      ast/state.sml
+   in
+      structure gs_76 = State
+   end
+   local
+      ast/sizes-sig.sml
+   in
+      signature gs_77 = SIZES
+   end
+   local
+      signature SIZES = gs_77
+      ast/sizes.sml
+   in
+      structure gs_78 = Sizes
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Sizes = gs_78
+      structure State = gs_76
+      structure Tables = gs_74
+      structure TextIO = gs_0
+      structure Tidtab = gs_73
+      ast/parse-to-ast-sig.sml
+   in
+      signature gs_79 = PARSE_TO_AST
+   end
+   local
+      open l4
+   in
+      structure gs_80 = ListPair
+   end
+   local
+      open l4
+   in
+      structure gs_81 = Option
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Error = gs_25
+      structure ParseTree = gs_32
+      structure Sizes = gs_78
+      structure State = gs_76
+      structure Tables = gs_74
+      structure Tidtab = gs_73
+      ast/build-ast-sig.sml
+   in
+      signature gs_82 = BUILD_AST
+   end
+   local
+      structure Ast = gs_69
+      structure ParseTree = gs_32
+      structure ParseTreeExt = gs_30
+      structure State = gs_76
+      ast/cnv-ext-sig.sml
+   in
+      signature gs_83 = CNVEXT
+   end
+   local
+      structure Ast = gs_69
+      signature CNVEXT = gs_83
+      structure ParseTree = gs_32
+      structure ParseTreeExt = gs_30
+      structure State = gs_76
+      ast/extensions/c/cnv-ext.sml
+   in
+      structure gs_84 = CnvExt
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Pid = gs_66
+      structure Symbol = gs_65
+      ast/simplify-assign-ops.sml
+   in
+      structure gs_85 = SimplifyAssignOps
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure Tables = gs_74
+      structure Tid = gs_63
+      ast/pp/pp-ast-sig.sml
+   in
+      signature gs_86 = PP_AST
+   end
+   local
+      open l4
+   in
+      structure gs_87 = Int32
+   end
+   local
+      structure Ast = gs_69
+      structure Int = gs_10
+      structure Int32 = gs_87
+      structure LargeInt = gs_29
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure Pid = gs_66
+      structure Real = gs_38
+      structure String = gs_15
+      structure Symbol = gs_65
+      structure Tables = gs_74
+      structure TextIO = gs_0
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      ast/pp/pp-lib.sml
+   in
+      structure gs_88 = PPLib
+   end
+   local
+      structure Ast = gs_69
+      structure AstExt = gs_67
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure Tables = gs_74
+      ast/pp/pp-ast-ext-sig.sml
+   in
+      signature gs_89 = PPASTEXT
+   end
+   local
+      signature PPASTEXT = gs_89
+      ast/extensions/c/pp-ast-ext-fn.sml
+   in
+      functor gs_90 = PPAstExtFn
+   end
+   local
+      structure Ast = gs_69
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure Tables = gs_74
+      ast/pp/pp-ast-adornment-sig.sml
+   in
+      signature gs_91 = PPASTADORNMENT
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Int = gs_10
+      structure LargeInt = gs_29
+      structure List = gs_21
+      signature OLD_PRETTYPRINT = gs_22
+      structure OldPrettyPrint = gs_23
+      structure Option = gs_81
+      signature PPASTADORNMENT = gs_91
+      functor PPAstExtFn = gs_90
+      structure PPLib = gs_88
+      signature PP_AST = gs_86
+      structure Pid = gs_66
+      structure SourceMap = gs_17
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      ast/pp/pp-ast-fn.sml
+   in
+      functor gs_92 = PPAstFn
+   end
+   local
+      structure Ast = gs_69
+      signature PPASTADORNMENT = gs_91
+      functor PPAstFn = gs_92
+      ast/pp/pp-ast.sml
+   in
+      structure gs_93 = PPAst
+   end
+   local
+      structure Ast = gs_69
+      ast/ctype-eq.sml
+   in
+      structure gs_94 = CTypeEq
+   end
+   local
+      structure Ast = gs_69
+      structure Sizes = gs_78
+      structure Tables = gs_74
+      ast/sizeof-sig.sml
+   in
+      signature gs_95 = SIZEOF
+   end
+   local
+      structure Ast = gs_69
+      structure LargeInt = gs_29
+      structure Tables = gs_74
+      ast/type-util-sig.sml
+   in
+      signature gs_96 = TYPE_UTIL
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Config = gs_4
+      structure Int = gs_10
+      structure List = gs_21
+      structure PPAst = gs_93
+      structure PPLib = gs_88
+      structure Pid = gs_66
+      structure Symbol = gs_65
+      signature TYPE_UTIL = gs_96
+      structure Tables = gs_74
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      ast/type-util.sml
+   in
+      structure gs_97 = TypeUtil
+   end
+   local
+      structure Ast = gs_69
+      functor BinaryMapFn = gs_71
+      structure Bindings = gs_70
+      structure Config = gs_4
+      structure Int = gs_10
+      structure LargeInt = gs_29
+      structure List = gs_21
+      structure Pid = gs_66
+      signature SIZEOF = gs_95
+      structure Sizes = gs_78
+      structure Tables = gs_74
+      structure TextIO = gs_0
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      structure TypeUtil = gs_97
+      ast/sizeof.sml
+   in
+      structure gs_98 = Sizeof
+   end
+   local
+      structure ParseTree = gs_32
+      structure Real = gs_38
+      structure Tid = gs_63
+      ast/anonymous-structs.sml
+   in
+      structure gs_99 = AnonymousStructs
+      structure gs_100 = TyEq
+   end
+   local
+      structure Aid = gs_12
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure Tid = gs_63
+      ast/initializer-normalizer-sig.sml
+   in
+      signature gs_101 = INITIALIZER_NORMALIZER
+   end
+   local
+      structure Aid = gs_12
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      signature INITIALIZER_NORMALIZER = gs_101
+      structure LargeInt = gs_29
+      structure SourceMap = gs_17
+      structure String = gs_15
+      structure Tid = gs_63
+      ast/initializer-normalizer.sml
+   in
+      structure gs_102 = InitializerNormalizer
+   end
+   local
+      structure Aid = gs_12
+      structure Aidtab = gs_13
+      structure AnonymousStructs = gs_99
+      structure Ast = gs_69
+      signature BUILD_AST = gs_82
+      functor BinaryMapFn = gs_71
+      structure Bindings = gs_70
+      structure CTypeEq = gs_94
+      structure CnvExt = gs_84
+      structure Config = gs_4
+      structure Error = gs_25
+      structure InitializerNormalizer = gs_102
+      structure Int = gs_10
+      structure LargeInt = gs_29
+      structure List = gs_21
+      structure ListPair = gs_80
+      structure Option = gs_81
+      structure PPAst = gs_93
+      structure PPLib = gs_88
+      structure ParseTree = gs_32
+      structure ParseTreeExt = gs_30
+      structure Pid = gs_66
+      structure SimplifyAssignOps = gs_85
+      structure Sizeof = gs_98
+      structure Sizes = gs_78
+      structure SourceMap = gs_17
+      structure State = gs_76
+      structure String = gs_15
+      structure Symbol = gs_65
+      structure Tables = gs_74
+      structure TextIO = gs_0
+      structure Tid = gs_63
+      structure Tidtab = gs_73
+      structure TyEq = gs_100
+      structure TypeUtil = gs_97
+      structure Word = gs_7
+      ast/build-ast.sml
+   in
+      structure gs_103 = BuildAst
+   end
+   local
+      structure Ast = gs_69
+      structure Bindings = gs_70
+      structure BuildAst = gs_103
+      structure Error = gs_25
+      signature PARSE_TO_AST = gs_79
+      structure PPAst = gs_93
+      structure PPLib = gs_88
+      structure Parser = gs_61
+      structure Sizes = gs_78
+      structure State = gs_76
+      structure Tables = gs_74
+      structure TextIO = gs_0
+      structure Tidtab = gs_73
+      ast/parse-to-ast.sml
+   in
+      structure gs_104 = ParseToAst
+   end
+in
+   signature AST = gs_68
+   structure Aidtab = gs_13
+   structure Ast = gs_69
+   structure Bindings = gs_70
+   signature CONFIG = gs_3
+   structure Config = gs_4
+   signature PARSECONTROL = gs_2
+   signature PARSER = gs_33
+   signature PARSETREE = gs_31
+   signature PARSE_TO_AST = gs_79
+   structure PPAst = gs_93
+   signature PP_AST = gs_86
+   structure ParseToAst = gs_104
+   structure ParseTree = gs_32
+   structure Parser = gs_61
+   structure Pid = gs_66
+   signature SOURCE_MAP = gs_16
+   structure Sizeof = gs_98
+   structure Sizes = gs_78
+   structure SourceMap = gs_17
+   structure State = gs_76
+   structure Symbol = gs_65
+   signature TYPECHECKCONTROL = gs_1
+   structure Tables = gs_74
+   structure Tid = gs_63
+   structure Tidtab = gs_73
+   structure TypeUtil = gs_97
+   signature UID = gs_8
+   signature UIDTABIMP = gs_6
+end
+end
+
+end
diff --git a/src/parser/grammar/c.grm.sml b/src/parser/grammar/c.grm.sml
index 175f5c4..1404fd2 100644
--- a/src/parser/grammar/c.grm.sml
+++ b/src/parser/grammar/c.grm.sml
@@ -1658,7 +1658,7 @@ local
        fun f i =
             if i=numstates then g i
             else (Array.update(memo,i,SHIFT (STATE i)); f (i+1))
-          in f 0 handle Subscript => ()
+          in f 0 handle General.Subscript => ()
           end
 in
 val entry_to_action = fn 0 => ACCEPT | 1 => ERROR | j => Array.sub(memo,(j-2))
@@ -1668,7 +1668,7 @@ val actionRows=string_to_table(string_to_pairlist_default(T,entry_to_action),act
 val actionRowNumbers = string_to_list actionRowNumbers
 val actionT = let val actionRowLookUp=
 let val a=Array.fromList(actionRows) in fn i=>Array.sub(a,i) end
-in Array.fromList(map actionRowLookUp actionRowNumbers)
+in Array.fromList(List.map actionRowLookUp actionRowNumbers)
 end
 in LrTable.mkLrTable {actions=actionT,gotos=gotoT,numRules=numrules,
 numStates=numstates,initialState=STATE 0}
diff --git a/src/parser/parse-tree-sig.sml b/src/parser/parse-tree-sig.sml
index 4ac8151..a50f78c 100644
--- a/src/parser/parse-tree-sig.sml
+++ b/src/parser/parse-tree-sig.sml
@@ -27,8 +27,10 @@ sig
     | ModAssign | XorAssign | OrAssign | AndAssign
     | LshiftAssign | RshiftAssign 
     | Uplus 
-    | SizeofType of ctype
-    | OperatorExt of operatorExt
+    | SizeofType of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | OperatorExt of
+        ParseTreeExt.operatorExt
 
   and expression
     = EmptyExpr
@@ -40,10 +42,17 @@ sig
     | Binop of operator * expression * expression
     | QuestionColon of expression * expression * expression
     | Call of expression * expression list
-    | Cast of ctype * expression
+    | Cast of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+        expression
     | InitList of expression list
     | MARKexpression of (SourceMap.location * expression)
-    | ExprExt of expressionExt
+    | ExprExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.expressionExt
 
   and specifier
     = Void
@@ -60,11 +69,15 @@ sig
     | Wholenum
     | Saturate
     | Nonsaturate
-    | Array of expression * ctype
-    | Pointer of ctype
+    | Array of
+        expression *
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | Pointer of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
     | Function of
-        {retType : ctype,  
-	 params : (decltype * declarator) list}
+        {retType : (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         params : ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+                   declarator) list}
     | Enum of
         {tagOpt : string option,
 	 enumerators : (string * expression) list,
@@ -72,13 +85,19 @@ sig
     | Struct of
         {isStruct : bool,   (* struct or union; true => struct *)
 	 tagOpt : string option,  (* optional tag *)
-	 members: (ctype * (declarator * expression) list) list} (* member specs *)
+         members: ((* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+                   (declarator * expression) list) list} (* member specs *)
     | TypedefName of string
     | StructTag of
 	{isStruct : bool,   (* ??? *)
 	 name : string}
     | EnumTag of string 
-    | SpecExt of specifierExt
+    | SpecExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.specifierExt
 
   and declarator  (* constructor suffix: "Decr" *)
     = EmptyDecr
@@ -87,9 +106,17 @@ sig
     | ArrayDecr of declarator * expression
     | PointerDecr of declarator
     | QualDecr of qualifier * declarator
-    | FuncDecr of declarator * (decltype * declarator) list
+    | FuncDecr of
+        declarator *
+        ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+         declarator) list
     | MARKdeclarator of (SourceMap.location * declarator)
-    | DecrExt of declaratorExt
+    | DecrExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.declaratorExt
 
   (* supports extensions of C in which expressions contain statements *)
   and statement
@@ -110,25 +137,42 @@ sig
     | IfThenElse of expression * statement * statement
     | Switch of expression * statement
     | MARKstatement of (SourceMap.location * statement)
-    | StatExt of statementExt
+    | StatExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.statementExt
 
   and declaration
-    = Declaration of decltype * (declarator * expression) list
+    = Declaration of
+        (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+        (declarator * expression) list
     | MARKdeclaration of (SourceMap.location * declaration)
-    | DeclarationExt of declarationExt
+    | DeclarationExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.declarationExt
 
   (* the top-level constructs in a translation unit (i.e. source file) *)
   and externalDecl
     = ExternalDecl of declaration
     | FunctionDef of
-       {retType : decltype,      (* return type *)
+       {retType : (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},      (* return type *)
 	funDecr : declarator,   (* function name declarator *)
         krParams : declaration list, (* K&R-style parameter declarations *)
         body : statement}        (* function body *)
     | MARKexternalDecl of (SourceMap.location * externalDecl)
-    | ExternalDeclExt of externalDeclExt
+    | ExternalDeclExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.externalDeclExt
 
-  withtype ctype =
+  type ctype =
       {qualifiers : qualifier list,
        specifiers : specifier list}
   and decltype =
@@ -137,7 +181,7 @@ sig
        storage : storage list}
 
   (* extension types for basic constructs *)
-  and externalDeclExt = 
+  type externalDeclExt =
       (specifier, declarator, ctype, decltype, operator, expression, statement)
       ParseTreeExt.externalDeclExt
   and declarationExt = 
diff --git a/src/parser/parse-tree.sml b/src/parser/parse-tree.sml
index 160eb5e..b62b955 100644
--- a/src/parser/parse-tree.sml
+++ b/src/parser/parse-tree.sml
@@ -23,8 +23,10 @@ struct
     | ModAssign | XorAssign | OrAssign | AndAssign
     | LshiftAssign | RshiftAssign 
     | Uplus 
-    | SizeofType of ctype
-    | OperatorExt of operatorExt
+    | SizeofType of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | OperatorExt of
+        ParseTreeExt.operatorExt
 
   and expression
     = EmptyExpr
@@ -36,10 +38,17 @@ struct
     | Binop of operator * expression * expression
     | QuestionColon of expression * expression * expression
     | Call of expression * expression list
-    | Cast of ctype * expression
+    | Cast of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+	expression
     | InitList of expression list
     | MARKexpression of (SourceMap.location * expression)
-    | ExprExt of expressionExt
+    | ExprExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.expressionExt
 
   and specifier
     = Void
@@ -56,11 +65,15 @@ struct
     | Wholenum
     | Saturate
     | Nonsaturate
-    | Array of expression * ctype
-    | Pointer of ctype
+    | Array of
+        expression *
+	(* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | Pointer of
+	(* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
     | Function of
-        {retType : ctype, 
-	 params : (decltype * declarator) list}
+        {retType : (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 params : ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+		   declarator) list}
     | Enum of
         {tagOpt : string option,
 	 enumerators : (string * expression) list,
@@ -68,13 +81,19 @@ struct
     | Struct of
         {isStruct : bool,   (* struct or union; true => struct *)
 	 tagOpt : string option,  (* optional tag *)
-	 members: (ctype * (declarator * expression) list) list} (* member specs *)
+	 members: ((* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+		   (declarator * expression) list) list} (* member specs *)
     | TypedefName of string
     | StructTag of
 	{isStruct : bool,   (* ??? *)
 	 name : string}
     | EnumTag of string 
-    | SpecExt of specifierExt
+    | SpecExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+        ParseTreeExt.specifierExt
 
   and declarator  (* constructor suffix: "Decr" *)
     = EmptyDecr
@@ -83,9 +102,17 @@ struct
     | ArrayDecr of declarator * expression
     | PointerDecr of declarator
     | QualDecr of qualifier * declarator
-    | FuncDecr of declarator * (decltype * declarator) list
+    | FuncDecr of
+        declarator *
+	((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+	 declarator) list
     | MARKdeclarator of (SourceMap.location * declarator)
-    | DecrExt of declaratorExt
+    | DecrExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.declaratorExt
 
   (* supports extensions of C in which expressions contain statements *)
   and statement
@@ -106,24 +133,41 @@ struct
     | IfThenElse of expression * statement * statement
     | Switch of expression * statement
     | MARKstatement of (SourceMap.location * statement)
-    | StatExt of statementExt
+    | StatExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.statementExt
 
   and declaration
-    = Declaration of decltype * (declarator * expression) list
+    = Declaration of
+        (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+        (declarator * expression) list
     | MARKdeclaration of (SourceMap.location * declaration)
-    | DeclarationExt of declarationExt
+    | DeclarationExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.declarationExt
 
   and externalDecl
     = ExternalDecl of declaration
     | FunctionDef of (* record? *)
-       {retType : decltype,      (* return type *)
+       {retType : (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},      (* return type *)
 	funDecr : declarator,   (* function name declarator *)
         krParams : declaration list, (* K&R-style parameter declarations *)
         body : statement}        (* function body *)
     | MARKexternalDecl of (SourceMap.location * externalDecl)
-    | ExternalDeclExt of externalDeclExt
+    | ExternalDeclExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.externalDeclExt
 
-  withtype ctype =
+  type ctype =
            {qualifiers : qualifier list,
 	    specifiers : specifier list}
   and decltype =
@@ -131,7 +175,7 @@ struct
        specifiers : specifier list,
        storage : storage list}      
 
-  and externalDeclExt = 
+  type externalDeclExt =
       (specifier, declarator, ctype, decltype, operator, expression, statement)
       ParseTreeExt.externalDeclExt
   and declarationExt = 
