diff --git a/README.mlton b/README.mlton
new file mode 100644
index 0000000..f1b9b15
--- /dev/null
+++ b/README.mlton
@@ -0,0 +1,8 @@
+The following changes were made to the ckit Library, in addition to
+deriving the `.mlb` file from the `.cm` file:
+
+* `ast/pp/pp-ast-adornment-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `ast/pp/pp-ast-ext-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `ast/type-util-sig.sml` (modified): Rewrote use of `signature` in `local`.
+* `parser/parse-tree-sig.sml` (modified): Rewrote use of (sequential) `withtype` in signature.
+* `parser/parse-tree.sml` (modified): Rewrote use of (sequential) `withtype`.
diff --git a/ckit-lib.mlb b/ckit-lib.mlb
new file mode 100644
index 0000000..abc601f
--- /dev/null
+++ b/ckit-lib.mlb
@@ -0,0 +1 @@
+src/ckit-lib.mlb
diff --git a/src/ast/pp/pp-ast-adornment-sig.sml b/src/ast/pp/pp-ast-adornment-sig.sml
index a7b937b..ab6e0a9 100644
--- a/src/ast/pp/pp-ast-adornment-sig.sml
+++ b/src/ast/pp/pp-ast-adornment-sig.sml
@@ -1,14 +1,14 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a pp =  Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
 
   type ('aidinfo,'a,'b) adornment_pp = ('aidinfo -> 'a) -> 'aidinfo -> 'b
-in
+(* in *)
 signature PPASTADORNMENT = sig
   type aidinfo
   val ppExpressionAdornment: (aidinfo,Ast.coreExpression pp,Ast.expression pp) adornment_pp
   val ppStatementAdornment : (aidinfo,Ast.coreStatement pp,Ast.statement pp) adornment_pp
   val ppExternalDeclAdornment: (aidinfo,Ast.coreExternalDecl pp,Ast.externalDecl pp) adornment_pp
 end
-end
+(* end *)
diff --git a/src/ast/pp/pp-ast-ext-sig.sml b/src/ast/pp/pp-ast-ext-sig.sml
index 4169fc7..c7291ab 100644
--- a/src/ast/pp/pp-ast-ext-sig.sml
+++ b/src/ast/pp/pp-ast-ext-sig.sml
@@ -1,13 +1,13 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a pp =  Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
   type ('a, 'aidinfo) ppExt =
        (('aidinfo -> Ast.expression pp) * ('aidinfo -> Ast.statement pp) *
 	('aidinfo -> Ast.binop pp) * ('aidinfo -> Ast.unop pp))
        -> 'aidinfo
        -> Tables.tidtab -> OldPrettyPrint.ppstream -> 'a -> unit
-in
+(* in *)
 
 signature PPASTEXT = sig
   type aidinfo
@@ -24,4 +24,4 @@ signature PPASTEXT = sig
        aidinfo) ppExt
 end
 
-end
+(* end *)
diff --git a/src/ast/type-util-sig.sml b/src/ast/type-util-sig.sml
index b03260b..a262146 100644
--- a/src/ast/type-util-sig.sml
+++ b/src/ast/type-util-sig.sml
@@ -1,10 +1,10 @@
 (* Copyright (c) 1998 by Lucent Technologies *)
 
-local 
+(* local *)
   type 'a type_util      = Tables.tidtab -> Ast.ctype               -> 'a 
   type 'a type_mem_util  = Tables.tidtab -> Ast.ctype * Ast.member  -> 'a 
   type 'a type_type_util = Tables.tidtab -> Ast.ctype * Ast.ctype -> 'a 
-in
+(* in *)
 
 signature TYPE_UTIL =
 sig
@@ -145,4 +145,4 @@ sig
 
 end (* signature TYPE_UTIL *)
 
-end (* local *)
+(* end (* local *) *)
diff --git a/src/ckit-lib.mlb b/src/ckit-lib.mlb
new file mode 100644
index 0000000..7266fc5
--- /dev/null
+++ b/src/ckit-lib.mlb
@@ -0,0 +1,1525 @@
+
+ann
+  "nonexhaustiveBind warn" "nonexhaustiveMatch warn"
+  "redundantBind warn" "redundantMatch warn"
+  "sequenceNonUnit ignore"
+  "warnUnused false" "forceUsed"
+  "allowOrPats true" "allowSigWithtype true" "allowVectorExpsAndPats true"
+in
+
+local
+  basis l11 =
+    bas
+      (* $/basis.cm ====> *) $(SML_LIB)/basis/basis.mlb $(SML_LIB)/basis/sml-nj.mlb $(SML_LIB)/basis/unsafe.mlb
+    end
+  basis l6 =
+    bas
+      (* $/smlnj-lib.cm ====> *) $(SML_LIB)/smlnj-lib/Util/smlnj-lib.mlb
+    end
+  basis l74 =
+    bas
+      (* $/pp-lib.cm ====> *) $(SML_LIB)/smlnj-lib/PP/pp-lib.mlb
+    end
+  basis l227 =
+    bas
+      (* $/ml-yacc-lib.cm ====> *) $(SML_LIB)/mlyacc-lib/mlyacc-lib.mlb
+    end
+in
+local
+   $(SML_LIB)/basis/pervasive.mlb
+   local
+      open l6
+   in
+      structure gs_0 = Format
+   end
+   local
+      open l11
+   in
+      structure gs_1 = TextIO
+   end
+   local
+      variants/type-check-control-sig.sml
+   in
+      signature gs_2 = TYPECHECKCONTROL
+   end
+   local
+      variants/parse-control-sig.sml
+   in
+      signature gs_3 = PARSECONTROL
+   end
+   local
+      signature PARSECONTROL = gs_3
+      signature TYPECHECKCONTROL = gs_2
+      variants/config-sig.sml
+   in
+      signature gs_4 = CONFIG
+   end
+   local
+      signature CONFIG = gs_4
+      signature PARSECONTROL = gs_3
+      signature TYPECHECKCONTROL = gs_2
+      structure TextIO = gs_1
+      variants/ansic/config.sml
+   in
+      structure gs_5 = Config
+   end
+   local
+      open l11
+   in
+      structure gs_6 = Char
+      structure gs_7 = CharArray
+      structure gs_8 = CharVector
+      structure gs_9 = FixedInt
+      structure gs_10 = General
+      structure gs_11 = Int
+      structure gs_12 = Int32
+      structure gs_13 = Int64
+      structure gs_14 = IntInf
+      structure gs_15 = LargeInt
+      structure gs_16 = LargeReal
+      structure gs_17 = LargeWord
+      structure gs_18 = OS
+      structure gs_19 = Position
+      structure gs_20 = Real
+      structure gs_21 = Real64
+      structure gs_22 = RealArray
+      structure gs_23 = RealArraySlice
+      structure gs_24 = RealVector
+      structure gs_25 = RealVectorSlice
+      structure gs_26 = SMLofNJ
+      structure gs_27 = Socket
+      structure gs_28 = String
+      structure gs_29 = Substring
+      structure gs_30 = SysWord
+      structure gs_31 = Time
+      structure gs_32 = Word
+      structure gs_33 = Word32
+      structure gs_34 = Word64
+      structure gs_35 = Word8
+   end
+   local
+      parser/util/sourcemap-sig.sml
+   in
+      signature gs_36 = SOURCE_MAP
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Config = gs_5
+      structure FixedInt = gs_9
+      structure Format = gs_0
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      signature SOURCE_MAP = gs_36
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/util/sourcemap.sml
+   in
+      structure gs_37 = SourceMap
+   end
+   local
+      open l11
+   in
+      structure gs_38 = Option
+   end
+   local
+      open l11
+   in
+      structure gs_39 = List
+   end
+   local
+      open l74
+   in
+      functor gs_40 = PPStreamFn
+   end
+   local
+      open l74
+   in
+      structure gs_41 = StringToken
+   end
+   local
+      open l11
+   in
+      structure gs_42 = StringCvt
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      functor PPStreamFn = gs_40
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure StringCvt = gs_42
+      structure StringToken = gs_41
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/util/old-pp.sml
+   in
+      signature gs_43 = OLD_PRETTYPRINT
+      structure gs_44 = OldPrettyPrint
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/uid-sig.sml
+   in
+      signature gs_45 = UID
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      signature UID = gs_45
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/uid-fn.sml
+   in
+      functor gs_46 = UidFn
+   end
+   local
+      signature UID = gs_45
+      functor UidFn = gs_46
+      ast/aid.sml
+   in
+      structure gs_47 = Aid
+   end
+   local
+      signature UID = gs_45
+      functor UidFn = gs_46
+      ast/tid.sml
+   in
+      structure gs_48 = Tid
+   end
+   local
+      open l6
+   in
+      structure gs_49 = HashString
+   end
+   local
+      structure Tid = gs_48
+      ast/symbol-sig.sml
+   in
+      signature gs_50 = SYMBOL
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure HashString = gs_49
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      signature SYMBOL = gs_50
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/symbol.sml
+   in
+      structure gs_51 = Symbol
+   end
+   local
+      ast/extensions/c/ast-ext.sml
+   in
+      structure gs_52 = AstExt
+   end
+   local
+      signature UID = gs_45
+      functor UidFn = gs_46
+      ast/pid.sml
+   in
+      structure gs_53 = Pid
+   end
+   local
+      structure Aid = gs_47
+      structure AstExt = gs_52
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/ast-sig.sml
+   in
+      signature gs_54 = AST
+   end
+   local
+      signature AST = gs_54
+      structure Aid = gs_47
+      structure AstExt = gs_52
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/ast.sml
+   in
+      structure gs_55 = Ast
+   end
+   local
+      structure Ast = gs_55
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/bindings.sml
+   in
+      structure gs_56 = Bindings
+   end
+   local
+      open l6
+   in
+      functor gs_57 = HashTableFn
+   end
+   local
+      ast/uidtabimp-sig.sml
+   in
+      signature gs_58 = UIDTABIMP
+   end
+   local
+      functor HashTableFn = gs_57
+      signature UID = gs_45
+      signature UIDTABIMP = gs_58
+      ast/uidtabimp-fn.sml
+   in
+      functor gs_59 = UidtabImpFn
+   end
+   local
+      structure Aid = gs_47
+      functor UidtabImpFn = gs_59
+      ast/aidtab.sml
+   in
+      structure gs_60 = Aidtab
+   end
+   local
+      structure Tid = gs_48
+      functor UidtabImpFn = gs_59
+      ast/tidtab.sml
+   in
+      structure gs_61 = Tidtab
+   end
+   local
+      structure Aidtab = gs_60
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Tidtab = gs_61
+      ast/tables.sml
+   in
+      structure gs_62 = Tables
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      signature OLD_PRETTYPRINT = gs_43
+      structure OldPrettyPrint = gs_44
+      structure Tables = gs_62
+      structure Tid = gs_48
+      ast/pp/pp-ast-sig.sml
+   in
+      signature gs_63 = PP_AST
+   end
+   local
+      structure Ast = gs_55
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      signature OLD_PRETTYPRINT = gs_43
+      structure OS = gs_18
+      structure OldPrettyPrint = gs_44
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure TextIO = gs_1
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/pp/pp-lib.sml
+   in
+      structure gs_64 = PPLib
+   end
+   local
+      structure Ast = gs_55
+      structure AstExt = gs_52
+      signature OLD_PRETTYPRINT = gs_43
+      structure OldPrettyPrint = gs_44
+      structure Tables = gs_62
+      ast/pp/pp-ast-ext-sig.sml
+   in
+      signature gs_65 = PPASTEXT
+   end
+   local
+      signature PPASTEXT = gs_65
+      ast/extensions/c/pp-ast-ext-fn.sml
+   in
+      functor gs_66 = PPAstExtFn
+   end
+   local
+      structure Ast = gs_55
+      signature OLD_PRETTYPRINT = gs_43
+      structure OldPrettyPrint = gs_44
+      structure Tables = gs_62
+      ast/pp/pp-ast-adornment-sig.sml
+   in
+      signature gs_67 = PPASTADORNMENT
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      signature OLD_PRETTYPRINT = gs_43
+      structure OS = gs_18
+      structure OldPrettyPrint = gs_44
+      structure Option = gs_38
+      signature PPASTADORNMENT = gs_67
+      functor PPAstExtFn = gs_66
+      structure PPLib = gs_64
+      signature PP_AST = gs_63
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/pp/pp-ast-fn.sml
+   in
+      functor gs_68 = PPAstFn
+   end
+   local
+      structure Ast = gs_55
+      signature PPASTADORNMENT = gs_67
+      functor PPAstFn = gs_68
+      ast/pp/pp-ast.sml
+   in
+      structure gs_69 = PPAst
+   end
+   local
+      open l6
+   in
+      functor gs_70 = BinaryMapFn
+   end
+   local
+      ast/sizes-sig.sml
+   in
+      signature gs_71 = SIZES
+   end
+   local
+      signature SIZES = gs_71
+      ast/sizes.sml
+   in
+      structure gs_72 = Sizes
+   end
+   local
+      structure Ast = gs_55
+      structure Sizes = gs_72
+      structure Tables = gs_62
+      ast/sizeof-sig.sml
+   in
+      signature gs_73 = SIZEOF
+   end
+   local
+      structure Ast = gs_55
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/type-util-sig.sml
+   in
+      signature gs_74 = TYPE_UTIL
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Config = gs_5
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      structure PPAst = gs_69
+      structure PPLib = gs_64
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      signature TYPE_UTIL = gs_74
+      structure Tables = gs_62
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/type-util.sml
+   in
+      structure gs_75 = TypeUtil
+   end
+   local
+      structure Ast = gs_55
+      functor BinaryMapFn = gs_70
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Config = gs_5
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      signature SIZEOF = gs_73
+      structure SMLofNJ = gs_26
+      structure Sizes = gs_72
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure TextIO = gs_1
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure TypeUtil = gs_75
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/sizeof.sml
+   in
+      structure gs_76 = Sizeof
+   end
+   local
+      structure Format = gs_0
+      signature OLD_PRETTYPRINT = gs_43
+      structure OldPrettyPrint = gs_44
+      structure SourceMap = gs_37
+      structure TextIO = gs_1
+      parser/util/error-sig.sml
+   in
+      signature gs_77 = ERROR
+   end
+   local
+      signature ERROR = gs_77
+      structure Format = gs_0
+      signature OLD_PRETTYPRINT = gs_43
+      structure OldPrettyPrint = gs_44
+      structure SourceMap = gs_37
+      structure TextIO = gs_1
+      parser/util/error.sml
+   in
+      structure gs_78 = Error
+   end
+   local
+      parser/extensions/c/parse-tree-ext.sml
+   in
+      structure gs_79 = ParseTreeExt
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure ParseTreeExt = gs_79
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/parse-tree-sig.sml
+   in
+      signature gs_80 = PARSETREE
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      signature PARSETREE = gs_80
+      structure ParseTreeExt = gs_79
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/parse-tree.sml
+   in
+      structure gs_81 = ParseTree
+   end
+   local
+      structure Error = gs_78
+      structure ParseTree = gs_81
+      parser/parser-sig.sml
+   in
+      signature gs_82 = PARSER
+   end
+   local
+      open l227
+   in
+      functor gs_83 = Join
+      functor gs_84 = JoinWithArg
+   end
+   local
+      open l227
+   in
+      structure gs_85 = LrParser
+   end
+   local
+      open l11
+   in
+      structure gs_86 = IO
+   end
+   local
+      open l11
+   in
+      structure gs_87 = TextPrimIO
+   end
+   local
+      open l11
+   in
+      structure gs_88 = Vector
+   end
+   local
+      open l6
+   in
+      structure gs_89 = Atom
+   end
+   local
+      open l6
+   in
+      structure gs_90 = AtomTable
+   end
+   local
+      structure Atom = gs_89
+      structure AtomTable = gs_90
+      structure Config = gs_5
+      parser/grammar/tdefs.sml
+   in
+      signature gs_91 = TYPEDEFS
+      structure gs_92 = TypeDefs
+   end
+   local
+      open l227
+   in
+      signature gs_93 = ARG_LEXER
+      signature gs_94 = ARG_PARSER
+      signature gs_95 = LEXER
+      signature gs_96 = LR_PARSER
+      signature gs_97 = LR_TABLE
+      signature gs_98 = PARSER
+      signature gs_99 = PARSER_DATA
+      signature gs_100 = STREAM
+      signature gs_101 = TOKEN
+   end
+   local
+      signature ARG_LEXER = gs_93
+      signature ARG_PARSER = gs_94
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      signature LEXER = gs_95
+      signature LR_PARSER = gs_96
+      signature LR_TABLE = gs_97
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      signature PARSER = gs_98
+      signature PARSER_DATA = gs_99
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      signature STREAM = gs_100
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      signature TOKEN = gs_101
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/grammar/c.grm.sig
+   in
+      signature gs_102 = C_LRVALS
+      signature gs_103 = C_TOKENS
+   end
+   local
+      structure Atom = gs_89
+      structure AtomTable = gs_90
+      signature C_LRVALS = gs_102
+      signature C_TOKENS = gs_103
+      structure Config = gs_5
+      signature TYPEDEFS = gs_91
+      structure TypeDefs = gs_92
+      parser/grammar/tokentable.sml
+   in
+      signature gs_104 = TOKENTABLE
+      functor gs_105 = TokenTable
+   end
+   local
+      signature C_LRVALS = gs_102
+      signature C_TOKENS = gs_103
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure IO = gs_86
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure StringCvt = gs_42
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      signature TOKENTABLE = gs_104
+      structure TextIO = gs_1
+      structure TextPrimIO = gs_87
+      structure Time = gs_31
+      functor TokenTable = gs_105
+      structure Vector = gs_88
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/grammar/c.lex.sml
+   in
+      functor gs_106 = CLexFun
+   end
+   local
+      open l11
+   in
+      structure gs_107 = Array
+   end
+   local
+      signature ARG_LEXER = gs_93
+      signature ARG_PARSER = gs_94
+      structure Array = gs_107
+      signature C_LRVALS = gs_102
+      signature C_TOKENS = gs_103
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Error = gs_78
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      signature LEXER = gs_95
+      signature LR_PARSER = gs_96
+      signature LR_TABLE = gs_97
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      signature PARSER = gs_98
+      signature PARSER_DATA = gs_99
+      structure ParseTree = gs_81
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      signature STREAM = gs_100
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      signature TOKEN = gs_101
+      signature TYPEDEFS = gs_91
+      structure Time = gs_31
+      structure TypeDefs = gs_92
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      parser/grammar/c.grm.sml
+   in
+      functor gs_108 = LrValsFun
+   end
+   local
+      functor CLexFun = gs_106
+      structure Error = gs_78
+      functor Join = gs_83
+      functor JoinWithArg = gs_84
+      structure LrParser = gs_85
+      functor LrValsFun = gs_108
+      signature PARSER = gs_82
+      structure SourceMap = gs_37
+      signature TOKENTABLE = gs_104
+      signature TYPEDEFS = gs_91
+      structure TextIO = gs_1
+      functor TokenTable = gs_105
+      structure TypeDefs = gs_92
+      parser/parser.sml
+   in
+      structure gs_109 = Parser
+   end
+   local
+      open l6
+   in
+      signature gs_110 = ORD_MAP
+   end
+   local
+      structure Aid = gs_47
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Error = gs_78
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      signature ORD_MAP = gs_110
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/state-sig.sml
+   in
+      signature gs_111 = STATE
+   end
+   local
+      structure Aid = gs_47
+      structure Aidtab = gs_60
+      structure Ast = gs_55
+      functor BinaryMapFn = gs_70
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure Error = gs_78
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure OS = gs_18
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      signature STATE = gs_111
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/state.sml
+   in
+      structure gs_112 = State
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Sizes = gs_72
+      structure State = gs_112
+      structure Tables = gs_62
+      structure TextIO = gs_1
+      structure Tidtab = gs_61
+      ast/parse-to-ast-sig.sml
+   in
+      signature gs_113 = PARSE_TO_AST
+   end
+   local
+      open l11
+   in
+      structure gs_114 = ListPair
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Error = gs_78
+      structure ParseTree = gs_81
+      structure Sizes = gs_72
+      structure State = gs_112
+      structure Tables = gs_62
+      structure Tidtab = gs_61
+      ast/build-ast-sig.sml
+   in
+      signature gs_115 = BUILD_AST
+   end
+   local
+      structure Aid = gs_47
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Tid = gs_48
+      ast/initializer-normalizer-sig.sml
+   in
+      signature gs_116 = INITIALIZER_NORMALIZER
+   end
+   local
+      structure Aid = gs_47
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      signature INITIALIZER_NORMALIZER = gs_116
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/initializer-normalizer.sml
+   in
+      structure gs_117 = InitializerNormalizer
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure Pid = gs_53
+      structure Symbol = gs_51
+      ast/simplify-assign-ops.sml
+   in
+      structure gs_118 = SimplifyAssignOps
+   end
+   local
+      structure Ast = gs_55
+      structure ParseTree = gs_81
+      structure ParseTreeExt = gs_79
+      structure State = gs_112
+      ast/cnv-ext-sig.sml
+   in
+      signature gs_119 = CNVEXT
+   end
+   local
+      structure Ast = gs_55
+      signature CNVEXT = gs_119
+      structure ParseTree = gs_81
+      structure ParseTreeExt = gs_79
+      structure State = gs_112
+      ast/extensions/c/cnv-ext.sml
+   in
+      structure gs_120 = CnvExt
+   end
+   local
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure OS = gs_18
+      structure ParseTree = gs_81
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure Socket = gs_27
+      structure String = gs_28
+      structure Substring = gs_29
+      structure SysWord = gs_30
+      structure Tid = gs_48
+      structure Time = gs_31
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/anonymous-structs.sml
+   in
+      structure gs_121 = AnonymousStructs
+      structure gs_122 = TyEq
+   end
+   local
+      structure Ast = gs_55
+      ast/ctype-eq.sml
+   in
+      structure gs_123 = CTypeEq
+   end
+   local
+      structure Aid = gs_47
+      structure Aidtab = gs_60
+      structure AnonymousStructs = gs_121
+      structure Ast = gs_55
+      signature BUILD_AST = gs_115
+      functor BinaryMapFn = gs_70
+      structure Bindings = gs_56
+      structure CTypeEq = gs_123
+      structure Char = gs_6
+      structure CharArray = gs_7
+      structure CharVector = gs_8
+      structure CnvExt = gs_120
+      structure Config = gs_5
+      structure Error = gs_78
+      structure FixedInt = gs_9
+      structure General = gs_10
+      structure InitializerNormalizer = gs_117
+      structure Int = gs_11
+      structure Int32 = gs_12
+      structure Int64 = gs_13
+      structure IntInf = gs_14
+      structure LargeInt = gs_15
+      structure LargeReal = gs_16
+      structure LargeWord = gs_17
+      structure List = gs_39
+      structure ListPair = gs_114
+      structure OS = gs_18
+      structure Option = gs_38
+      structure PPAst = gs_69
+      structure PPLib = gs_64
+      structure ParseTree = gs_81
+      structure ParseTreeExt = gs_79
+      structure Pid = gs_53
+      structure Position = gs_19
+      structure Real = gs_20
+      structure Real64 = gs_21
+      structure RealArray = gs_22
+      structure RealArraySlice = gs_23
+      structure RealVector = gs_24
+      structure RealVectorSlice = gs_25
+      structure SMLofNJ = gs_26
+      structure SimplifyAssignOps = gs_118
+      structure Sizeof = gs_76
+      structure Sizes = gs_72
+      structure Socket = gs_27
+      structure SourceMap = gs_37
+      structure State = gs_112
+      structure String = gs_28
+      structure Substring = gs_29
+      structure Symbol = gs_51
+      structure SysWord = gs_30
+      structure Tables = gs_62
+      structure TextIO = gs_1
+      structure Tid = gs_48
+      structure Tidtab = gs_61
+      structure Time = gs_31
+      structure TyEq = gs_122
+      structure TypeUtil = gs_75
+      structure Word = gs_32
+      structure Word32 = gs_33
+      structure Word64 = gs_34
+      structure Word8 = gs_35
+      ast/build-ast.sml
+   in
+      structure gs_124 = BuildAst
+   end
+   local
+      structure Ast = gs_55
+      structure Bindings = gs_56
+      structure BuildAst = gs_124
+      structure Error = gs_78
+      signature PARSE_TO_AST = gs_113
+      structure PPAst = gs_69
+      structure PPLib = gs_64
+      structure Parser = gs_109
+      structure Sizes = gs_72
+      structure State = gs_112
+      structure Tables = gs_62
+      structure TextIO = gs_1
+      structure Tidtab = gs_61
+      ast/parse-to-ast.sml
+   in
+      structure gs_125 = ParseToAst
+   end
+in
+   signature AST = gs_54
+   structure Aidtab = gs_60
+   structure Ast = gs_55
+   structure Bindings = gs_56
+   signature CONFIG = gs_4
+   structure Config = gs_5
+   signature PARSECONTROL = gs_3
+   signature PARSER = gs_82
+   signature PARSETREE = gs_80
+   signature PARSE_TO_AST = gs_113
+   structure PPAst = gs_69
+   signature PP_AST = gs_63
+   structure ParseToAst = gs_125
+   structure ParseTree = gs_81
+   structure Parser = gs_109
+   structure Pid = gs_53
+   signature SOURCE_MAP = gs_36
+   structure Sizeof = gs_76
+   structure Sizes = gs_72
+   structure SourceMap = gs_37
+   structure State = gs_112
+   structure Symbol = gs_51
+   signature TYPECHECKCONTROL = gs_2
+   structure Tables = gs_62
+   structure Tid = gs_48
+   structure Tidtab = gs_61
+   structure TypeUtil = gs_75
+   signature UID = gs_45
+   signature UIDTABIMP = gs_58
+end
+end
+
+end
diff --git a/src/parser/parse-tree-sig.sml b/src/parser/parse-tree-sig.sml
index 4ac8151..a50f78c 100644
--- a/src/parser/parse-tree-sig.sml
+++ b/src/parser/parse-tree-sig.sml
@@ -27,8 +27,10 @@ sig
     | ModAssign | XorAssign | OrAssign | AndAssign
     | LshiftAssign | RshiftAssign 
     | Uplus 
-    | SizeofType of ctype
-    | OperatorExt of operatorExt
+    | SizeofType of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | OperatorExt of
+        ParseTreeExt.operatorExt
 
   and expression
     = EmptyExpr
@@ -40,10 +42,17 @@ sig
     | Binop of operator * expression * expression
     | QuestionColon of expression * expression * expression
     | Call of expression * expression list
-    | Cast of ctype * expression
+    | Cast of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+        expression
     | InitList of expression list
     | MARKexpression of (SourceMap.location * expression)
-    | ExprExt of expressionExt
+    | ExprExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.expressionExt
 
   and specifier
     = Void
@@ -60,11 +69,15 @@ sig
     | Wholenum
     | Saturate
     | Nonsaturate
-    | Array of expression * ctype
-    | Pointer of ctype
+    | Array of
+        expression *
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | Pointer of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
     | Function of
-        {retType : ctype,  
-	 params : (decltype * declarator) list}
+        {retType : (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         params : ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+                   declarator) list}
     | Enum of
         {tagOpt : string option,
 	 enumerators : (string * expression) list,
@@ -72,13 +85,19 @@ sig
     | Struct of
         {isStruct : bool,   (* struct or union; true => struct *)
 	 tagOpt : string option,  (* optional tag *)
-	 members: (ctype * (declarator * expression) list) list} (* member specs *)
+         members: ((* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+                   (declarator * expression) list) list} (* member specs *)
     | TypedefName of string
     | StructTag of
 	{isStruct : bool,   (* ??? *)
 	 name : string}
     | EnumTag of string 
-    | SpecExt of specifierExt
+    | SpecExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.specifierExt
 
   and declarator  (* constructor suffix: "Decr" *)
     = EmptyDecr
@@ -87,9 +106,17 @@ sig
     | ArrayDecr of declarator * expression
     | PointerDecr of declarator
     | QualDecr of qualifier * declarator
-    | FuncDecr of declarator * (decltype * declarator) list
+    | FuncDecr of
+        declarator *
+        ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+         declarator) list
     | MARKdeclarator of (SourceMap.location * declarator)
-    | DecrExt of declaratorExt
+    | DecrExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.declaratorExt
 
   (* supports extensions of C in which expressions contain statements *)
   and statement
@@ -110,25 +137,42 @@ sig
     | IfThenElse of expression * statement * statement
     | Switch of expression * statement
     | MARKstatement of (SourceMap.location * statement)
-    | StatExt of statementExt
+    | StatExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.statementExt
 
   and declaration
-    = Declaration of decltype * (declarator * expression) list
+    = Declaration of
+        (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+        (declarator * expression) list
     | MARKdeclaration of (SourceMap.location * declaration)
-    | DeclarationExt of declarationExt
+    | DeclarationExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.declarationExt
 
   (* the top-level constructs in a translation unit (i.e. source file) *)
   and externalDecl
     = ExternalDecl of declaration
     | FunctionDef of
-       {retType : decltype,      (* return type *)
+       {retType : (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},      (* return type *)
 	funDecr : declarator,   (* function name declarator *)
         krParams : declaration list, (* K&R-style parameter declarations *)
         body : statement}        (* function body *)
     | MARKexternalDecl of (SourceMap.location * externalDecl)
-    | ExternalDeclExt of externalDeclExt
+    | ExternalDeclExt of
+        (specifier, declarator,
+         (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+         (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+         operator, expression, statement)
+        ParseTreeExt.externalDeclExt
 
-  withtype ctype =
+  type ctype =
       {qualifiers : qualifier list,
        specifiers : specifier list}
   and decltype =
@@ -137,7 +181,7 @@ sig
        storage : storage list}
 
   (* extension types for basic constructs *)
-  and externalDeclExt = 
+  type externalDeclExt =
       (specifier, declarator, ctype, decltype, operator, expression, statement)
       ParseTreeExt.externalDeclExt
   and declarationExt = 
diff --git a/src/parser/parse-tree.sml b/src/parser/parse-tree.sml
index 160eb5e..b62b955 100644
--- a/src/parser/parse-tree.sml
+++ b/src/parser/parse-tree.sml
@@ -23,8 +23,10 @@ struct
     | ModAssign | XorAssign | OrAssign | AndAssign
     | LshiftAssign | RshiftAssign 
     | Uplus 
-    | SizeofType of ctype
-    | OperatorExt of operatorExt
+    | SizeofType of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | OperatorExt of
+        ParseTreeExt.operatorExt
 
   and expression
     = EmptyExpr
@@ -36,10 +38,17 @@ struct
     | Binop of operator * expression * expression
     | QuestionColon of expression * expression * expression
     | Call of expression * expression list
-    | Cast of ctype * expression
+    | Cast of
+        (* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+	expression
     | InitList of expression list
     | MARKexpression of (SourceMap.location * expression)
-    | ExprExt of expressionExt
+    | ExprExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.expressionExt
 
   and specifier
     = Void
@@ -56,11 +65,15 @@ struct
     | Wholenum
     | Saturate
     | Nonsaturate
-    | Array of expression * ctype
-    | Pointer of ctype
+    | Array of
+        expression *
+	(* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
+    | Pointer of
+	(* ctype *) {qualifiers : qualifier list, specifiers : specifier list}
     | Function of
-        {retType : ctype, 
-	 params : (decltype * declarator) list}
+        {retType : (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 params : ((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+		   declarator) list}
     | Enum of
         {tagOpt : string option,
 	 enumerators : (string * expression) list,
@@ -68,13 +81,19 @@ struct
     | Struct of
         {isStruct : bool,   (* struct or union; true => struct *)
 	 tagOpt : string option,  (* optional tag *)
-	 members: (ctype * (declarator * expression) list) list} (* member specs *)
+	 members: ((* ctype *) {qualifiers : qualifier list, specifiers : specifier list} *
+		   (declarator * expression) list) list} (* member specs *)
     | TypedefName of string
     | StructTag of
 	{isStruct : bool,   (* ??? *)
 	 name : string}
     | EnumTag of string 
-    | SpecExt of specifierExt
+    | SpecExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+        ParseTreeExt.specifierExt
 
   and declarator  (* constructor suffix: "Decr" *)
     = EmptyDecr
@@ -83,9 +102,17 @@ struct
     | ArrayDecr of declarator * expression
     | PointerDecr of declarator
     | QualDecr of qualifier * declarator
-    | FuncDecr of declarator * (decltype * declarator) list
+    | FuncDecr of
+        declarator *
+	((* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+	 declarator) list
     | MARKdeclarator of (SourceMap.location * declarator)
-    | DecrExt of declaratorExt
+    | DecrExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.declaratorExt
 
   (* supports extensions of C in which expressions contain statements *)
   and statement
@@ -106,24 +133,41 @@ struct
     | IfThenElse of expression * statement * statement
     | Switch of expression * statement
     | MARKstatement of (SourceMap.location * statement)
-    | StatExt of statementExt
+    | StatExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.statementExt
 
   and declaration
-    = Declaration of decltype * (declarator * expression) list
+    = Declaration of
+        (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list} *
+        (declarator * expression) list
     | MARKdeclaration of (SourceMap.location * declaration)
-    | DeclarationExt of declarationExt
+    | DeclarationExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.declarationExt
 
   and externalDecl
     = ExternalDecl of declaration
     | FunctionDef of (* record? *)
-       {retType : decltype,      (* return type *)
+       {retType : (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},      (* return type *)
 	funDecr : declarator,   (* function name declarator *)
         krParams : declaration list, (* K&R-style parameter declarations *)
         body : statement}        (* function body *)
     | MARKexternalDecl of (SourceMap.location * externalDecl)
-    | ExternalDeclExt of externalDeclExt
+    | ExternalDeclExt of
+        (specifier, declarator,
+	 (* ctype *) {qualifiers : qualifier list, specifiers : specifier list},
+	 (* decltype *) {qualifiers : qualifier list, specifiers : specifier list, storage : storage list},
+	 operator, expression, statement)
+	ParseTreeExt.externalDeclExt
 
-  withtype ctype =
+  type ctype =
            {qualifiers : qualifier list,
 	    specifiers : specifier list}
   and decltype =
@@ -131,7 +175,7 @@ struct
        specifiers : specifier list,
        storage : storage list}      
 
-  and externalDeclExt = 
+  type externalDeclExt =
       (specifier, declarator, ctype, decltype, operator, expression, statement)
       ParseTreeExt.externalDeclExt
   and declarationExt = 
