diff -Nur -x '*.orig' -x '*~' mlton/bin/mlton-script mlton.new/bin/mlton-script
--- mlton/bin/mlton-script	2009-11-01 18:37:00.000000000 +0100
+++ mlton.new/bin/mlton-script	2009-11-01 18:45:26.000000000 +0100
@@ -7,7 +7,14 @@
 dir=`dirname "$0"`
 lib=`cd "$dir/../lib" && pwd`
 eval `"$lib/platform"`
-gcc='gcc'
+case "$HOST_ARCH" in
+mips*)
+        gcc='gcc-4.4'
+;;
+*)
+        gcc='gcc'
+;;
+esac
 case "$HOST_OS" in
 mingw)
 	exe='.exe'
