diff --git a/lib/ml-lpt-lib.mlb b/lib/ml-lpt-lib.mlb
index 77fd389..828a82a 100644
--- a/lib/ml-lpt-lib.mlb
+++ b/lib/ml-lpt-lib.mlb
@@ -1,38 +1,151 @@
-(* ml-lpt-lib.mlb
- *
- * COPYRIGHT (c) 2008
- * John Reppy (http://www.cs.uchicago.edu/~jhr)
- * Aaron Turon (http://www.cs.uchicago.edu/~adrassi)
- * All rights reserved.
- *
- * MLB file for ml-lpt lib
- *)
-
-local
-
-  $(SML_LIB)/basis/basis.mlb
-  $(SML_LIB)/basis/sml-nj.mlb
-  $(SML_LIB)/smlnj-lib/Util/smlnj-lib.mlb
-
-  stream-pos.sml
-  ebnf.sml
-  antlr-lexer-sig.sml
-  antlr-tokens-sig.sml
-  repair.sml
-  wrapped-strm.sml
-  err-handler.sml
-  ulex-buffer.sml
 
+ann
+  "nonexhaustiveMatch warn" "redundantMatch warn"
+  "sequenceNonUnit ignore"
+  "warnUnused false" "forceUsed"
 in
 
-  signature ANTLR_LEXER
-  signature ANTLR_TOKENS
-
-  structure AntlrRepair
-  structure AntlrStreamPos
-  structure ULexBuffer
-
-  functor AntlrEBNF
-  functor AntlrErrHandler
+local
+  basis l8 =
+    bas
+      (* $/basis.cm ====> *) $(SML_LIB)/basis/basis.mlb $(SML_LIB)/basis/sml-nj.mlb
+    end
+  basis l62 =
+    bas
+      (* $/smlnj-lib.cm ====> *) $(SML_LIB)/smlnj-lib/Util/smlnj-lib.mlb
+    end
+in
+local
+   $(SML_LIB)/basis/pervasive.mlb
+   local
+      antlr-tokens-sig.sml
+   in
+      signature gs_0 = ANTLR_TOKENS
+   end
+   local
+      open l8
+   in
+      structure gs_1 = String
+   end
+   local
+      open l8
+   in
+      structure gs_2 = List
+   end
+   local
+      open l8
+   in
+      structure gs_3 = Position
+   end
+   local
+      open l8
+   in
+      structure gs_4 = Int
+   end
+   local
+      structure Int = gs_4
+      structure Position = gs_3
+      structure String = gs_1
+      stream-pos.sml
+   in
+      structure gs_5 = AntlrStreamPos
+   end
+   local
+      structure AntlrStreamPos = gs_5
+      structure List = gs_2
+      structure String = gs_1
+      repair.sml
+   in
+      structure gs_6 = AntlrRepair
+   end
+   local
+      structure AntlrStreamPos = gs_5
+      structure List = gs_2
+      ebnf.sml
+   in
+      functor gs_7 = AntlrEBNF
+   end
+   local
+      structure AntlrStreamPos = gs_5
+      antlr-lexer-sig.sml
+   in
+      signature gs_8 = ANTLR_LEXER
+   end
+   local
+      open l8
+   in
+      structure gs_9 = SMLofNJ
+   end
+   local
+      open l8
+   in
+      structure gs_10 = ListPair
+   end
+   local
+      signature ANTLR_LEXER = gs_8
+      signature ANTLR_TOKENS = gs_0
+      structure AntlrRepair = gs_6
+      structure AntlrStreamPos = gs_5
+      structure Int = gs_4
+      structure ListPair = gs_10
+      structure String = gs_1
+      wrapped-strm.sml
+   in
+      functor gs_11 = AntlrWrappedStream
+   end
+   local
+      signature ANTLR_LEXER = gs_8
+      signature ANTLR_TOKENS = gs_0
+      structure AntlrRepair = gs_6
+      structure AntlrStreamPos = gs_5
+      functor AntlrWrappedStream = gs_11
+      structure Int = gs_4
+      structure List = gs_2
+      structure SMLofNJ = gs_9
+      err-handler.sml
+   in
+      functor gs_12 = AntlrErrHandler
+   end
+   local
+      open l8
+   in
+      structure gs_13 = Word
+   end
+   local
+      open l8
+   in
+      structure gs_14 = Substring
+   end
+   local
+      open l8
+   in
+      structure gs_15 = Char
+   end
+   local
+      structure AntlrStreamPos = gs_5
+      structure Char = gs_15
+      structure String = gs_1
+      structure Substring = gs_14
+      structure Word = gs_13
+      ulex-buffer.sml
+   in
+      structure gs_16 = ULexBuffer
+   end
+   local
+      open l62
+   in
+      structure gs_17 = UTF8
+   end
+in
+   signature ANTLR_LEXER = gs_8
+   signature ANTLR_TOKENS = gs_0
+   functor AntlrEBNF = gs_7
+   functor AntlrErrHandler = gs_12
+   structure AntlrRepair = gs_6
+   structure AntlrStreamPos = gs_5
+   structure ULexBuffer = gs_16
+   structure UTF8 = gs_17
+end
+end
 
 end
diff --git a/lib/mllpt-lib.mlb b/lib/mllpt-lib.mlb
new file mode 100644
index 0000000..0d6bf2a
--- /dev/null
+++ b/lib/mllpt-lib.mlb
@@ -0,0 +1 @@
+ml-lpt-lib.mlb
